src/core/meshcop/border_agent.cpp

```diff
@@ -317,6 +337,217 @@ template <> void BorderAgent::HandleTmf<kUriRelayRx>(Coap::Message &aMessage, co
+    SuccessOrExit(error = AppendTxtEntryTo(appender, "xa", Get<Mac::Mac>().GetExtAddress()));
+
+    state = GetStateBitmap();
+    SuccessOrExit(error = AppendTxtEntryTo(appender, "sb", BigEndian::HostSwap32(state.ToUint32())));
+
+    if (state.mThreadIfStatus == kThreadIfStatusActive)
+    {
+        SuccessOrExit(
+            error = AppendTxtEntryTo(appender, "pt",
+                                     BigEndian::HostSwap32(Get<Mle::MleRouter>().GetLeaderData().GetPartitionId())));
+        SuccessOrExit(error = AppendActiveTimestampTxtEntry(appender));
+    }
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+    SuccessOrExit(error = AppendBbrTxtEntry(state, appender));
+#endif
+#if OTBR_ENABLE_BORDER_ROUTING
+    SuccessOrExit(error = AppendOmrTxtEntry(appender));
+#endif
+
+    aLength = appender.GetAppendedLength();
+
+exit:
+    return error;
+}
+
+Error BorderAgent::MeshCopTxtEncoder::AppendActiveTimestampTxtEntry(Appender &aAppender)
+{
+    Error error = kErrorNone;
+
+    if (Get<MeshCoP::ActiveDatasetManager>().GetTimestamp().IsValid())
+    {
+        error = AppendTxtEntryTo(aAppender, "at", Get<MeshCoP::ActiveDatasetManager>().GetTimestamp());
+    }
+    else
+    {
+        LogWarn("Failed to get valid active timestamp");
+    }
+
+    return error;
+}
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+Error BorderAgent::MeshCopTxtEncoder::AppendBbrTxtEntry(StateBitmap aState, Appender &aAppender)
+{
+    Error             error      = kErrorNone;
+    const DomainName &domainName = Get<MeshCoP::NetworkNameManager>().GetDomainName();
+
+    if (aState.mBbrIsActive)
+    {
+        BackboneRouter::Config bbrConfig;
+
+        Get<BackboneRouter::Local>().GetConfig(bbrConfig);
+        SuccessOrExit(error = AppendTxtEntryTo(aAppender, "sq", bbrConfig.mSequenceNumber));
+        SuccessOrExit(error =
+                          AppendTxtEntryTo(aAppender, "bb", BigEndian::HostSwap16(BackboneRouter::kBackboneUdpPort)));
+    }
+
+    error = AppendTxtEntryTo(
+        aAppender, "dn",
+        NameData(domainName.GetAsCString(), StringLength(domainName.GetAsCString(), sizeof(domainName))));
+
+exit:
+    return error;
+}
+#endif // OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+Error BorderAgent::MeshCopTxtEncoder::AppendOmrTxtEntry(Appender &aAppender)
+{
+    Error                                         error = kErrorNone;
+    Ip6::Prefix                                   prefix;
+    BorderRouter::RoutingManager::RoutePreference preference;
+
+    if ((error = Get<BorderRouter::RoutingManager>().GetFavoredOmrPrefix(prefix, preference)) == kErrorNone)
+    {
+        uint8_t omrData[Ip6::NetworkPrefix::kSize + 1];
+        omrData[0] = prefix.mLength;
+        memcpy(omrData + 1, prefix.mPrefix.mFields.m8, (prefix.mLength + 7) / 8);
```

Comment:
Can use `Prefix::SizeForLength()` here instead of `(prefix.mLength + 7) / 8`?

Reply:
Or even better, `Ip6::Prefix` has `GetBytesSize()` which gives the number of bytes.